src/main/java/org/broadinstitute/consent/http/service/DatasetRegistrationService.java

```diff
@@ -0,0 +1,426 @@
+
+        dataUse.setMethodsResearch(Objects.nonNull(group.getNmds()) ? !group.getNmds() : null); // TODO: is this right?
+        dataUse.setCommercialUse(Objects.nonNull(group.getNpu()) ? !group.getNpu() : null);
+        dataUse.setOther(group.getOtherPrimary());
+        dataUse.setSecondaryOther(group.getOtherSecondary());
+        dataUse.setPopulationOriginsAncestry(group.getPoa());
+        dataUse.setPublicationResults(group.getPub());
+
+        return dataUse;
+    }
+
+    private List<FileStorageObject> uploadFiles(Map<String, FormDataBodyPart> files,
+                                                Map<String, BlobId> uploadedFileCache,
+                                                Integer consentGroupIdx,
+                                                User user) throws IOException {
+        List<FileStorageObject> consentGroupFSOs = new ArrayList<>();
+
+        if (files.containsKey("alternativeDataSharingPlan")) {
+            consentGroupFSOs.add(uploadFile(
+                                 files, uploadedFileCache, user,
+                                 "alternativeDataSharingPlan",
+                                 FileCategory.ALTERNATIVE_DATA_SHARING_PLAN));
+        }
+
+        if (files.containsKey("consentGroups["+consentGroupIdx+"].nihInstitutionalCertificationFile")) {
+            consentGroupFSOs.add(uploadFile(
+                    files, uploadedFileCache, user,
+                    "consentGroups["+ consentGroupIdx +"].nihInstitutionalCertificationFile",
+                    FileCategory.NIH_INSTITUTIONAL_CERTIFICATION));
+        }
+
+        return consentGroupFSOs;
+
+    }
+
+    private FileStorageObject uploadFile(Map<String, FormDataBodyPart> files,
+                                         Map<String, BlobId> uploadedFileCache,
+                                         User user,
+                                         String name,
+                                         FileCategory category) throws IOException {
+
+        FormDataBodyPart bodyPart = files.get(name);
+
+        if (!uploadedFileCache.containsKey(name)) {
+            BlobId id = uploadFile(bodyPart);
+            uploadedFileCache.put(name, id);
+        }
+
+        BlobId id = uploadedFileCache.get(name);
+
+        FileStorageObject fso = new FileStorageObject();
+        fso.setCategory(category);
+        fso.setFileName(bodyPart.getContentDisposition().getFileName());
+        fso.setMediaType(bodyPart.getMediaType().toString());
+        fso.setBlobId(id);
+        fso.setCreateUserId(user.getUserId());
+
+        return fso;
+    }
+
+
+    /**
+     * Extracts an individual field as a dataset property.
+     *
+     * @param name The human-readable name of the field
+     * @param schemaProp The schema property name (camelCase)
+     * @param type The type of the field, e.g. Boolean, String
+     * @param getField Lambda which gets the field's value
+     */
+    private record DatasetPropertyExtractor(
```

Comment:
It would be nice to extract this to a public record so it could be independently testable. As it stands now, all usages are private so testing needs to go through the service class.

Reply:
hmm, I agreed at first, but I have always been taught in classes that things shouldn't be exposed just for testing. I was also able to use argument captors to get into the nitty gritty of testing exactly what was sent to the dataset insert, and I was able to test dataset property creation then. It is a lot for that one method to test, but the method is basically "convert dataset registration into dataset insert files" and from that perspective, it makes sense to test it all. I'm ok if y'all both have a preference for breaking it up, though.

Reply:
> I have always been taught in classes that things shouldn't be exposed just for testing

TBH, I disagree with that, but it isn't `just for testing`, it's more about organization. Inner, private classes are good for small units of functionality but when there's any significant logic, it needs to stand on its own. This one is doing a lot of work that needs to be exercised properly. It's all really good stuff and there's a lot of potential for reuse. Imagine trying to do an insert with a V2 schema that subclassed V1 but with more fields ... there'd need to be some work done here to handle that and I'd want to make sure tests covered all cases. Probably not the best example, but it's the first to come to mind.

Reply:
That makes sense, actually. I'll make it public and add test cases for it.